### PR TITLE
Feature/logging

### PR DESCRIPTION
**Pull request by soccer-fan**
All tests pass - I have updated the ActionOption an abstract class to add an abstract method override for toString - all subclasses should override toString to describe their action in a manner that can be used for logging (ie "$unitInstanceName $actionOption").

I have also added a bit of logic to CombatSimulation to handle multiple rounds and fighting until one side is all dead.

I have set a simple end to end test in the WorldCreatorTest file, which basically just creates a world from the test jsons, simulates combat and asserts that one side has been wiped out at the end of combat.

I have left in the logger stuff in case we get sl4j working in the future,

### CODE REVIEW

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt [hunk 1]
```diff
@@ -18,16 +21,51 @@ class CombatSimulation(val world: World) {
...
+            if((unit.currentAttributes.getValue(Attribute.CURRENT_HEALTH) as NumericValue).value > 0) {
+                // check if unit alive rather than try removing dead units from list as concurrency risks
+                // caller needs to exception handle this in case CURRENT_HEALTH not provided in json
+                unit.tacticalActionStrategy
+                        .decideTurnActions(world, unit, unit.availableActionOptions)
+                        .let { actionsToExecute ->
+                            unit.executeActions(actionsToExecute)
+                        }
+                        .also {
+                            val deadUnits = world.findDead()
```
Comment (HSAR) [inline src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt:35]: You can avoid having to dedicated `val` statement by starting a function chain.
`world.findDead().let { deadUnits -> ... }.forEach { deadUnit -> ... }`

Comment (HSAR) [inline src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt:35]: You can avoid having to dedicated `val` statement by starting a function chain.
`world.findDead().let { deadUnits -> ... }.forEach { deadUnit -> ... }`

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt [hunk 2]
```diff
@@ -18,16 +21,51 @@ class CombatSimulation(val world: World) {
...
         }
     }
+
+    fun runSimulation() {
+        var roundNum = 1
+        while(world.enemyForces.size > 0 && world.friendlyForces.size > 0) {
+            printRoundStartStatus(roundNum)
+            try {
+                runRound()
+            } catch(e: Exception) {
```
Comment (HSAR) [inline src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt:52]: I appreciate the desire for safety but I don't agree with this, I think that our system is fairly closed and controlled and thus we should never see exceptions in normal operation (this is the correct use of exceptions for "exceptional behaviour").

That's a very roundabout way of saying that we should just let exceptions reach the user and terminate the run. It's not like it's super expensive to start another run.

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt [hunk 3]
```diff
@@ -18,16 +21,51 @@ class CombatSimulation(val world: World) {
...
+        if(world.enemyForces.size == 0) {
+            println("Party wins")
+        } else {
+            println("Enemies win")
+        }
+    }
+
+    private fun printRoundStartStatus(roundNum: Int) {
+        println("Start of combat round $roundNum")
+        println("Friendly units alive: ${world.friendlyForces.map{it.name}}")
```
Comment (HSAR) [inline src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt:67]: [Minor] You should be able to straight up print `${world.friendlyForces}`. It looks a bit more machine-y but I personally prefer it.

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/logic/ActionOption.kt [hunk 4]
```diff
@@ -21,6 +21,7 @@ import io.hsar.wh40k.combatsimulator.random.RollResult
 @JsonIgnoreProperties(value = ["actionType"])
 sealed class ActionOption {
     abstract val actionCost: ActionCost
+    abstract override fun toString(): String
```
Comment (HSAR) [inline src/main/kotlin/io/hsar/wh40k/combatsimulator/logic/ActionOption.kt:24]: [Minor] You should always be able to use the class name, which would be a bit brutal but should always provide something. I recommend a default implementation.

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/model/UnitInstance.kt [hunk 5]
```diff
@@ -20,12 +22,11 @@ class UnitInstance(
         val equipment: List<EquipmentItem>,
         val attackExecutor: AttackExecutor = AttackExecutor(), // open for test
         val startingAttributes: Map<Attribute, AttributeValue> =  // #TODO: Figure out whether this is good long-term solution
-                DEFAULT_ATTRIBUTES + equipment.map { it.modifiesAttributes }.sum(),
+                DEFAULT_ATTRIBUTES + equipment.map { it.modifiesAttributes }.sum()
+                        + Pair(Attribute.CURRENT_HEALTH, NumericValue(unit.stats.baseStats.getValue(BaseStat.MAX_HEALTH))),
```
Comment (HSAR) [inline src/main/kotlin/io/hsar/wh40k/combatsimulator/model/UnitInstance.kt:26]: Common syntactic sugar for creating `Pair("foo", "bar")` is `"foo" to "bar"`.

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/model/UnitInstance.kt [hunk 6]
```diff
@@ -80,13 +83,15 @@ class UnitInstance(
...
+        println("$name took $damage damage")
         when(val health = currentAttributes.getValue(CURRENT_HEALTH)) {
             is NumericValue -> currentAttributes[CURRENT_HEALTH] = NumericValue(health.value - damage)
             else -> throw IllegalStateException("Current health ought to be a NumericValue")
         }
     }
 
-    companion object {
+    companion object : Loggable {
+        val log = logger()
```
Comment (HSAR) [inline src/main/kotlin/io/hsar/wh40k/combatsimulator/model/UnitInstance.kt:94]: Unused logger?

**Review by HSAR — DISMISSED** [review]
Nothing really that I'd insist on having done before merge.

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/model/UnitInstance.kt (hunk 6 above)
Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/model/UnitInstance.kt:94]: Will be used in the future if we can get logging config working

Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/model/UnitInstance.kt:94]: Can remove for now if you like?

Comment (HSAR) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/model/UnitInstance.kt:94]: Yeah, remove it for now. It's a bit misleading.

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt (hunk 3 above)
Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt:67]: It's very verbose - contrast the below:
Friendly units alive: [io.hsar.wh40k.combatsimulator.model.UnitInstance@7573e12f, io.hsar.wh40k.combatsimulator.model.UnitInstance@51b1a8f6, io.hsar.wh40k.combatsimulator.model.UnitInstance@2a389173, io.hsar.wh40k.combatsimulator.model.UnitInstance@213860b8]
Enemy units alive: [banditRanged1, banditRanged2, banditRanged3, banditRanged4, banditMelee1, banditMelee2]

Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt:67]: If we are going to print to stdout at least for the time being as our main output method then I think making it shorter and more human-readable is a better option. It also makes it easier to tie up the units taking actions with the units in the JSON file as you can see the names used in the JSON.

Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt:67]: Open to ideas?

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/logic/ActionOption.kt (hunk 4 above)
Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/logic/ActionOption.kt:24]: As with comments elsewhere in the PR ref output to user, I think a modicum of polish to make output easily readable and digestible is OK. I can change it if you want though?

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/model/UnitInstance.kt (hunk 5 above)
Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/model/UnitInstance.kt:26]: Updated

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt (hunk 2 above)
Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt:52]: Removed.

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt (hunk 1 above)
Comment (markyMitch) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt:35]: Updated.

File: src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt (hunk 3 above)
Comment (HSAR) [reply src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt]: That's a bit of a pain, I guess `UnitInstance` isn't a data class. Yeah, no worries then. If you overrode `toString()` to use `name` that would be a bit more general purpose, that's what I would do but the effect is basically the same.

**Review by HSAR — APPROVED** [review]

**Pull request merged**